idp/adfs/adfs.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package adfs is an identity provider that authenticates with an ADFS
+// service.
+package adfs
+
+import (
+	oidc "github.com/coreos/go-oidc"
+	"gopkg.in/errgo.v1"
+
+	"github.com/canonical/candid/idp"
+	"github.com/canonical/candid/idp/openid"
+)
+
+func init() {
+	idp.Register("adfs", func(unmarshal func(interface{}) error) (idp.IdentityProvider, error) {
+		var p Params
+		if err := unmarshal(&p); err != nil {
+			return nil, errgo.Notef(err, "cannot unmarshal adfs parameters")
+		}
+		if p.URL == "" {
+			return nil, errgo.Newf("url not specified")
+		}
+		if p.ClientID == "" {
+			return nil, errgo.Newf("client-id not specified")
+		}
+		if p.ClientSecret == "" {
+			return nil, errgo.Newf("client-secret not specified")
+		}
+		return NewIdentityProvider(p), nil
+	})
+}
+
+type Params struct {
+	// Name is the name that will be given to the identity provider.
+	Name string `yaml:"name"`
+
+	// Description is the description that will be used with the
+	// identity provider. If this is not set then Name will be used.
+	Description string `yaml:"description"`
+
+	// Icon contains the URL or path of an icon.
+	Icon string `yaml:"icon"`
+
+	// Domain is the domain with which all identities created by this
+	// identity provider will be tagged (not including the @ separator).
+	Domain string `yaml:"domain"`
+
+	// URL is the URL of the Active Directory Federation Services
+	// instance that is used to provide identities. OpenID Connect
+	// discovery will be run on this URL to determine the required
+	// service parameters.
+	URL string `yaml:"url"`
+
+	// ClientID contains the Application Id for the application
+	// registered at
+	// https://console.developers.google.com/apis/credentials.
+	ClientID string `yaml:"client-id"`
+
+	// ClientSecret contains a password type Application Secret for
+	// the application as generated on
+	// https://console.developers.google.com/apis/credentials.
+	ClientSecret string `yaml:"client-secret"`
+
+	// Hidden is set if the IDP should be hidden from interactive
+	// prompts.
+	Hidden bool `yaml:"hidden"`
+}
+
+// NewIdentityProvider creates a google identity provider with the
```

Comment:
nit: ADFS instead of google?

Reply:
Thanks, I found a couple of others I've fixed too.